core/src/toga/window.py

```diff
@@ -681,43 +682,28 @@ def save_file_dialog(
         file_types: list[str] | None = None,
         on_result: DialogResultHandler[Path | None] | None = None,
     ) -> Dialog:
-        """Prompt the user for a location to save a file.
-
-        This dialog is not currently supported on Android or iOS.
-
-        **This is an asynchronous method**. If you invoke this method in synchronous
-        context, it will show the dialog, but will return *immediately*. The object
-        returned by this method can be awaited to obtain the result of the dialog.
-
-        :param title: The title of the dialog window
-        :param suggested_filename: A default filename
-        :param file_types: The allowed filename extensions, without leading dots. If not
-            provided, any extension will be allowed.
-        :param on_result: **DEPRECATED** ``await`` the return value of this method.
-        :returns: An awaitable Dialog object. The Dialog object returns a path object
-            for the selected file location, or ``None`` if the user cancelled the save
-            operation.
-        """
-        dialog = Dialog(
+        """**DEPRECATED** - use ``await self.dialog(SaveFileDialog())``"""
+        ######################################################################
+        # 2024-06: Backwards compatibility
+        ######################################################################
+        warnings.warn(
+            "save_file_dialog(...) has been deprecated; use dialog(toga.SaveFileDialog(...))",
+            DeprecationWarning,
+        )
+        ######################################################################
+        # End Backwards compatibility
+        ######################################################################
+        result = Dialog(
             self,
             on_result=wrapped_handler(self, on_result) if on_result else None,
         )
-        # Convert suggested filename to a path (if it isn't already),
-        # and break it into a filename and a directory
-        suggested_path = Path(suggested_filename)
-        initial_directory: Path | None = suggested_path.parent
-        if initial_directory == Path("."):
-            initial_directory = None
-        filename = suggested_path.name
-
-        self.factory.dialogs.SaveFileDialog(
-            dialog,
+        result.dialog = dialogs.SaveFileDialog(
             title,
-            filename=filename,
-            initial_directory=initial_directory,
+            suggested_filename=suggested_filename,
             file_types=file_types,
         )
-        return dialog
+        result.dialog._impl.show(self, result)
+        return result
 
     @overload
     def open_file_dialog(
```

Comment:
I understand these overloads are supposed to help with automated type inference, but I don't think it's worth the huge mess they make in the documentation. 

Suggest removing them, for all the dialog types, and NOT replicating them in the new API.

Reply:
Agreed the `@overload` declarations make a mess of the documentation. I preserved them on the basis that they were pre-existing, used mostly to ensure deprecated arguments are type safe, and pre-existing code might be relying on the specific overloading type hints to keep IDEs happy; but if an IDE starts complaining, then that's one more thing to make users to remove use of a deprecated API.

Definitely have no intention of replicating them in the new API.